query-parser/src/find.rs

```diff
@@ -0,0 +1,134 @@
+///!
+///! ```clojure
+///! {:find [?y]
+///!  :in [$]
+///!  :where [[?x :foaf/knows ?y]]}
+///! ```
+///!
+///! We parse by expanding the array format into four parts, treating them as the four
+///! parts of the map.
+
+extern crate edn;
+extern crate mentat_query;
+
+use std::collections::BTreeMap;
+
+use self::edn::Value::PlainSymbol;
+use self::mentat_query::{FindSpec, SrcVar, Variable};
+
+use super::error::{FindParseError, FindParseResult};
+use super::util::{values_to_variables, vec_to_keyword_map};
+
+fn parse_find_parts(find: &[edn::Value],
+                    ins: Option<&[edn::Value]>,
+                    with: Option<&[edn::Value]>,
+                    wheres: &[edn::Value]) -> FindParseResult {
+    // :find must be an array of plain var symbols (?foo), pull expressions, and aggregates.
+    // For now we only support variables and the annotations necessary to declare which
+    // flavor of :find we want:
+    //     ?x ?y ?z       = FindRel
+    //     [?x ...]       = FindColl
+    //     ?x .           = FindScalar
+    //     [?x ?y ?z]     = FindTuple
+    //
+    // :in must be an array of sources ($), rules (%), and vars (?). For now we only support the
+    // default source. :in can be omitted, in which case the default is equivalent to `:in $`.
+    // TODO: process `ins`.
+    let source = SrcVar::DefaultSrc;
+
+    // :with is an array of variables. This is simple, so we don't use a parser.
+    let with_vars = with.map(values_to_variables);
+
+    //
+    // :wheres is a whole datastructure.
+
+    Ok(FindSpec::FindRel(vec!()))
+}
+
+fn parse_find_map(map: BTreeMap<edn::Keyword, Vec<edn::Value>>) -> FindParseResult {
+    // Eagerly awaiting `const fn`.
+    let kw_find = edn::Keyword::new("find");
+    let kw_in = edn::Keyword::new("in");
+    let kw_with = edn::Keyword::new("with");
+    let kw_where = edn::Keyword::new("where");
+
+    // Oh, if only we had `guard`.
+    if let Some(find) = map.get(&kw_find) {
+        if let Some(wheres) = map.get(&kw_where) {
+            return parse_find_parts(find,
+                                    map.get(&kw_in).map(|x| x.as_slice()),
+                                    map.get(&kw_with).map(|x| x.as_slice()),
+                                    wheres);
+        } else {
+            return Err(FindParseError::MissingField(kw_where));
+        }
+    } else {
+        return Err(FindParseError::MissingField(kw_find));
+    }
+}
+
+fn parse_find_edn_map(map: BTreeMap<edn::Value, edn::Value>) -> FindParseResult {
+    // Every key must be a Keyword. Every value must be a Vec.
+    let mut m = BTreeMap::new();
+
+    if map.is_empty() {
+        return parse_find_map(m);
+    }
+
+    for (k, v) in map {
+        if let edn::Value::Keyword(kw) = k {
+            if let edn::Value::Vector(vec) = v {
+                m.insert(kw, vec);
+                continue;
+            } else {
+                return Err(FindParseError::InvalidInput(v));
```

Comment:
These errors are going to be so hard to debug.  I see two alternatives:

1. Assume that initial consumers will be comfortable inspecting and debugging Mentat itself; iterate on the error messages over time as confusion arises.
2.  Parse (probably using `combine`) the data structures directly, investing in error reporting in the parser combinator layer to help people not hate their lives.

Both are investments.